pkg/controller/handlers/mcpcatalog/mcpcatalog.go

```diff
@@ -273,3 +278,114 @@ func (h *Handler) SetUpDefaultMCPCatalog(ctx context.Context, c client.Client) e
 
 	return nil
 }
+
+// DeleteUnauthorizedMCPServers is a handler that deletes MCP servers that are no longer authorized to exist.
+// This can happen whenever AccessControlRules change.
+// It does not delete MCPServerInstances, since those have a delete ref to their MCPServer, and will be deleted automatically.
+func (h *Handler) DeleteUnauthorizedMCPServers(req router.Request, _ router.Response) error {
+	// List AccessControlRules so that this handler gets triggered any time one of them changes.
+	if err := req.List(&v1.AccessControlRuleList{}, &client.ListOptions{
+		Namespace: system.DefaultNamespace,
+	}); err != nil {
+		return fmt.Errorf("failed to list access control rules: %w", err)
+	}
+
+	var mcpServers v1.MCPServerList
+	if err := req.List(&mcpServers, &client.ListOptions{
+		Namespace: system.DefaultNamespace,
+	}); err != nil {
+		return fmt.Errorf("failed to list MCP servers: %w", err)
+	}
+
+	// Iterate through each MCPServer and make sure it is still allowed to exist.
+	for _, server := range mcpServers.Items {
+		if server.Spec.ToolReferenceName != "" || server.Spec.ThreadName != "" || server.Spec.SharedWithinMCPCatalogName != "" {
+			// For legacy gptscript tools, project-scoped servers, and multi-user servers created by the admin, we don't need to check them.
+			continue
+		}
+
+		user, err := h.gatewayClient.UserByID(req.Ctx, server.Spec.UserID)
+		if err != nil {
+			return fmt.Errorf("failed to get user %s: %w", server.Spec.UserID, err)
+		}
```

Comment:
Does the server get cleaned up by another controller if the user no longer exists? If not, should we clean it up here.

Reply:
Good call. I need to update the UserDelete controller route to delete the servers that belong to the user. I'll take care of that shortly.

Reply:
This is fixed now